group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -65,191 +94,233 @@ public String name() {
-                    unassignedPartitionsForTopic.add(i);
-                }
+        Map<String, MemberAssignment> assignments = new HashMap<>((int) ((groupSpec.memberIds().size() / 0.75f) + 1));
+
+        for (String memberId : memberIds) {
+            MemberSubscription subs = groupSpec.memberSubscription(memberId);
+            Map<Uuid, Set<Integer>> assignment = new HashMap<>(subs.subscribedTopicIds().size());
+            for (Uuid topicId : subs.subscribedTopicIds()) {
+                TopicMetadata metadata = topics.get(topicId);
+                metadata.maybeComputeQuota();
+                addPartitionsToAssignment(metadata, assignment);
             }
+            assignments.put(memberId, new MemberAssignmentImpl(assignment));
+        }
+
+        return new GroupAssignment(assignments);
+    }
 
-            // Step 4 and Step 5
-            // Account for the extra partitions if necessary and increase the required quota by 1.
-            // If remaining > 0 after increasing the required quota, assign the remaining number of partitions from the unassigned partitions list.
-            int unassignedPartitionsListStartPointer = 0;
-            for (MemberWithRemainingAssignments pair : potentiallyUnfilledMembers) {
-                String memberId = pair.memberId;
-                int remaining = pair.remaining;
-                if (numMembersWithExtraPartition > 0) {
-                    remaining++;
-                    numMembersWithExtraPartition--;
-                }
-                if (remaining > 0) {
-                    List<Integer> partitionsToAssign = unassignedPartitionsForTopic
-                        .subList(unassignedPartitionsListStartPointer, unassignedPartitionsListStartPointer + remaining);
-                    unassignedPartitionsListStartPointer += remaining;
-                    newTargetAssignment.computeIfAbsent(memberId, k -> new MemberAssignmentImpl(new HashMap<>()))
-                        .partitions()
-                        .computeIfAbsent(topicId, k -> new HashSet<>())
-                        .addAll(partitionsToAssign);
-                }
+    /**
+     * Sorts the member Ids in the group based on their instance Id if present, otherwise by member Id.
+     * This is done to ensure that the relative ordering of members doesn't change with static members
+     * thus resulting in a sticky assignment.
+     *
+     * @param groupSpec     The group specification containing the member information.
+     * @return a sorted list of member Ids.
+     */
+    private List<String> sortMemberIds(
+        GroupSpec groupSpec
+    ) {
+        List<String> sortedMemberIds = new ArrayList<>(groupSpec.memberIds());
+        Map<String, Optional<String>> instanceIdCache = new HashMap<>();
+
+        for (String memberId : sortedMemberIds) {
+            instanceIdCache.put(memberId, groupSpec.memberSubscription(memberId).instanceId());
+        }
```

Comment:
I wonder if this cache is necessary. Using it to look up the instance id or using the map backing `memberSubscription` seems to be equivalent. Does it provide a performance boost?

Reply:
yeah it provided a performance boost by a couple ms, like 8-9ms on the 10k members

Reply:
Ok. Let’s add a small comment about the reason then because it is not obvious.